vision/status_quo/nalgebra.md

```diff
@@ -0,0 +1,149 @@
+```
+
+To deal with the lack of generic const expressions, they add a trait for conversions from and to [`typenum`](https://crates.io/crates/typenum) for all `Const` up to size `127` ([source](https://github.com/dimforge/nalgebra/blob/39bb572557299a44093ea09daaff144fd6d9ea1f/src/base/dimension.rs#L273-L345)).
+
+Whenever they now need some computation using `Const<N>`, they convert it to type nums, evaluate the computation using the trait system, and then convert the result back to some `Const<M>`.
+
+## Disadvantages
+
+While this mostly works fine, there are some disadvantages.
+
+### Annoying `ToTypenum` bounds
+
+Most notably this adds a lot of unnecessary bounds, consider the following impl:
+
+```rust
+impl<T, const R1: usize, const C1: usize, const R2: usize, const C2: usize>
+    ReshapableStorage<T, Const<R1>, Const<C1>, Const<R2>, Const<C2>> for ArrayStorage<T, R1, C1>
+where
+    T: Scalar,
+    Const<R1>: ToTypenum,
+    Const<C1>: ToTypenum,
+    Const<R2>: ToTypenum,
+    Const<C2>: ToTypenum,
+    <Const<R1> as ToTypenum>::Typenum: Mul<<Const<C1> as ToTypenum>::Typenum>,
+    <Const<R2> as ToTypenum>::Typenum: Mul<
+        <Const<C2> as ToTypenum>::Typenum,
+        Output = typenum::Prod<
+            <Const<R1> as ToTypenum>::Typenum,
+            <Const<C1> as ToTypenum>::Typenum,
+        >,
+    >,
+{
+    type Output = ArrayStorage<T, R2, C2>;
+
+    fn reshape_generic(self, _: Const<R2>, _: Const<C2>) -> Self::Output {
+        unsafe {
+            let data: [[T; R2]; C2] = mem::transmute_copy(&self.0);
+            mem::forget(self.0);
+            ArrayStorage(data)
+        }
+    }
+}
+```
+
+### `ToTypenum` is only implemented up to fixed size
+
+That's annoying. ✨
+
+### Cannot use associated constants
+
+It is currently also not possible to have the size of a matrix depend on associated constants:
+```rust
+trait MyDimensions {
+   const ROWS: usize;
+   const COLS: usize;
+}
+
+fn foo<Dims: MyDimensions>() {
+    // Not possible!
+    let matrix: SMatrix<f64, Dims::ROWS, Dims::COLS> = SMatrix::zeros();
+}
+```
+While this can be avoided by going to back to `typenum` and using associated types, this adds a lot of unnecessary bounds and inpacts all of the code dealing with it.
+
+## Wishlist
+
+Ideally, `Matrix` could be changed to the following:
+
+```rust
+enum Dim {
+    Const(usize),
+    Dynamic,
+}
+
+struct Matrix<T, const R: Dim, const C: Dim, S> { ... }
+
+type SMatrix<T, const R: usize, const C: usize> =
+    Matrix<T, Dim::Const(R), Dim::Const(C), ArrayStorage<T, R, C>>;
+```
+
+For this to work well there have a bunch of requirements for const generics:
+
+### User-defined types as const parameter types
+
+We have to be able to use `Dim` as a const param type
+
+### Consider injective expressions to bind generic params
+
+With this change, `nalgebra` needs impls like the following
+
+```rust
+impl<T, const R: usize, const C: usize> for SMatrix<T, R, C> {
```

Comment:
Would this work if the type alias were expanded manually here?

Reply:
no, we still only see `ConstKind::Unevaluated` in the self type

Reply:
How is this case a problem then? For the later cases with exhaustive impls?

Reply:
because you can't write that impl
```rust
#![feature(adt_const_params, generic_const_exprs)]

struct Foo<const N: Option<usize>>;

impl<const N: usize> Foo<{ Some(N) }> {}
```
results in
```
error[[E0207]](https://doc.rust-lang.org/nightly/error-index.html#E0207): the const parameter `N` is not constrained by the impl trait, self type, or predicates
 [--> src/lib.rs:5:12
](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021#)  |
5 | impl<const N: usize> Foo<{ Some(N) }> {}
  |            ^ unconstrained const parameter
  |
  = note: expressions using a const parameter must map each value to a distinct output value
  = note: proving the result of expressions other than the parameter are unique is not supported
```